proposals/254-padding-negotiation.txt

```diff
@@ -436,42 +177,113 @@ increasing bin width allows the histograms to most accurately represent
+
+Clients will select which of the defined available padding machines to use
+based on the conditions that these machines specify. These conditions include:
+  * How many hops the circuit must be in order for the machine to apply
+  * If the machine requires vanguards to be enabled to apply
+  * The state the circuit must be in for machines to apply (building,
+    relay early cells remaining, opened, streams currently attached).
+  * If the circuit purpose matches a set of purposes for the machine.
+  * If the target hop of the machine supports circuit padding.
+
+Clients will only select machines whose conditions fully match given circuits.
+
+A machine is represented by a positive number that can be thought of as a "menu
+option" through the list of padding machines. The currently supported padding
+state machines are:
+
+        [1]: CIRCPAD_MACHINE_CIRC_SETUP
+
+             A padding machine that obscures the initial circuit setup in an
+             attempt to hide onion services.
+
+3.3. Machine Negotiation
+
+When a machine is selected, the client uses leaky-pipe delivery to send a
+RELAY_COMMAND_PADDING_NEGOTIATE to the target hop of the machine, using the
+following trunnel relay cell payload format:
+
+  /**
+   * This command tells the relay to alter its min and max netflow
+   * timeout range values, and send padding at that rate (resuming
+   * if stopped). */
+  struct circpad_negotiate {
+    u8 version IN [0];
+    u8 command IN [CIRCPAD_COMMAND_START, CIRCPAD_COMMAND_STOP];
 
-If the remove_tokens field is set to a non-zero value for a given
-state's histogram, then whenever a padding packet is sent, the
-corresponding histogram bin's token count is decremented by one.
+    /** Machine type is left unbounded because we can specify
+     * new machines in the consensus */
+    u8 machine_type;
+  };
 
-If a packet matching the current state's transition_reschedule_events
-bitmask arrives from the server before the chosen padding timer expires,
-then a token is removed from a non-empty bin corresponding to
-the delay since the last packet was sent, and the padding packet timer
-is re-sampled from the histogram.
+Upon receipt of a RELAY_COMMAND_PADDING_NEGOTIATE cell, the middle node sends
+a RELAY_COMMAND_PADDING_NEGOTIATED with the following format:
 
-The three enums for the remove_tokens field govern if we take the token
-out of the nearest lower non-empty bin, the nearest higher non-empty
-bin, or simply the closest non-empty bin.
+  /**
+   * This command tells the relay to alter its min and max netflow
+   * timeout range values, and send padding at that rate (resuming
+   * if stopped). */
+  struct circpad_negotiated {
+    u8 version IN [0];
+    u8 command IN [CIRCPAD_COMMAND_START, CIRCPAD_COMMAND_STOP];
+    u8 response IN [CIRCPAD_RESPONSE_OK, CIRCPAD_RESPONSE_ERR];
+
+    /** Machine type is left unbounded because we can specify
+     * new machines in the consensus */
+    u8 machine_type;
```

Comment:
What's the point of sending back the same machine_type? What kind of checks should be done here, if any?

Reply:
The machine_type is used to ensure that this STOP request is for the machine that is currently running, and to ensure that the reponse is similarly for the machine that was running.

This is to make sure we haven't replaced a machine with a different one in the time while we waited for the response to come back. (Machines can be replaced right away, but machine info is kept around until the response comes back so that we can count any padding in the interim as valid cells).

Reply:
Clarified this also in https://github.com/torproject/torspec/pull/41/commits/1dd7f1ff78fcb69121130cbd802b0b8b527ffc63